postgres/src/lib.rs

```diff
@@ -0,0 +1,693 @@
+                    history_segment: b"v1".to_vec()
+                })
+            );
+            Ok(())
+        })
+        .await
+    }
+
+    #[tokio::test]
+    /// When an add_version call specifies an incorrect `parent_version_id, it fails. This is
+    /// typically avoided by calling `get_client` beforehand, which (due to repeatable reads)
+    /// allows the caller to check the `latest_version_id` before calling `add_version`.
+    async fn add_version_mismatch() -> anyhow::Result<()> {
+        with_db(async |connection_string, db_client| {
+            let storage = PostgresStorage::new(connection_string).await?;
+            let client_id = make_client(&db_client).await?;
+            let latest_version_id = Uuid::new_v4();
+            set_client_latest_version_id(&db_client, client_id, latest_version_id).await?;
+
+            let mut txn = storage.txn(client_id).await?;
+            let version_id = Uuid::new_v4();
+            let parent_version_id = Uuid::new_v4(); // != latest_version_id
+            let res = txn
+                .add_version(version_id, parent_version_id, b"v1".to_vec())
+                .await;
+            assert!(res.is_err());
+            Ok(())
+        })
+        .await
+    }
+
+    #[tokio::test]
+    /// Adding versions to two different clients can proceed concurrently.
+    async fn add_version_no_conflict_different_clients() -> anyhow::Result<()> {
+        with_db(async |connection_string, db_client| {
+            let storage = PostgresStorage::new(connection_string).await?;
+
+            // Clients 1 and 2 do not interfere with each other; if these are the same client, then
+            // this will deadlock as one transaction waits for the other.
+            let client_id1 = make_client(&db_client).await?;
+            let mut txn1 = storage.txn(client_id1).await?;
+            let version_id1 = Uuid::new_v4();
+            txn1.add_version(version_id1, Uuid::nil(), b"v1".to_vec())
+                .await?;
+            assert_eq!(
+                txn1.get_version(version_id1).await?,
+                Some(Version {
+                    version_id: version_id1,
+                    parent_version_id: Uuid::nil(),
+                    history_segment: b"v1".to_vec()
+                })
+            );
+
+            let client_id2 = make_client(&db_client).await?;
+            let mut txn2 = storage.txn(client_id2).await?;
+            let version_id2 = Uuid::new_v4();
+            txn2.add_version(version_id2, Uuid::nil(), b"v2".to_vec())
+                .await?;
+            assert_eq!(
+                txn2.get_version(version_id2).await?,
+                Some(Version {
+                    version_id: version_id2,
+                    parent_version_id: Uuid::nil(),
+                    history_segment: b"v2".to_vec()
+                })
+            );
+
+            txn1.commit().await?;
+            txn2.commit().await?;
+
+            Ok(())
+        })
+        .await
+    }
```

Comment:
It's not clear to how this test would fail if the anticipated failure mode occurred. The assertions don't seem like the pertinent failure points unless I'm missing something. If they were the same client would we expect `txn2.add_version()` to block indefinitely and the test to fail by timing out? If the assertions aren't where we expect the failure then please add a comment at the relevant lines to make the purpose of this test clear.